core/src/Kokkos_CopyViews.hpp

```diff
@@ -1095,6 +1090,67 @@ inline void deep_copy(
   }
 }
 
+template <class DT, class... DP, class ST, class... SP>
+inline void deep_copy(
+    const View<DT, DP...>& dst, const View<ST, SP...>& src,
+    std::enable_if_t<(
+        std::is_void_v<typename ViewTraits<DT, DP...>::specialize> &&
+        std::is_void_v<typename ViewTraits<ST, SP...>::specialize> &&
+        !std::is_same_v<typename ViewTraits<DT, DP...>::value_type,
+                        typename ViewTraits<ST, SP...>::non_const_value_type> &&
+        (unsigned(ViewTraits<DT, DP...>::rank) == unsigned(0) &&
+         unsigned(ViewTraits<ST, SP...>::rank) == unsigned(0)))>* = nullptr) {
+  using dst_type = View<DT, DP...>;
+  using src_type = View<ST, SP...>;
+
+  static_assert(std::is_same_v<typename dst_type::value_type,
+                               typename dst_type::non_const_value_type>,
+                "deep_copy requires non-const destination type");
+
+  using dst_memory_space = typename dst_type::memory_space;
+  using src_memory_space = typename src_type::memory_space;
+
+  using exec_space = std::conditional_t<
+      Kokkos::SpaceAccessibility<typename dst_memory_space::execution_space,
+                                 src_memory_space>::accessible,
+      typename dst_memory_space::execution_space,
+      typename src_memory_space::execution_space>;
+
+  static_assert(
+      SpaceAccessibility<exec_space, dst_memory_space>::accessible &&
+          SpaceAccessibility<exec_space, src_memory_space>::accessible,
+      "deep_copy between different value types requires common accessibility "
+      "of Views");
+
+  if (Kokkos::Tools::Experimental::get_callbacks().begin_deep_copy != nullptr) {
+    Kokkos::Profiling::beginDeepCopy(
+        Kokkos::Profiling::make_space_handle(dst_memory_space::name()),
```

Comment:
It seems that this can be simplified to:
```suggestion
    Kokkos::Profiling::beginDeepCopy(
        Kokkos::Profiling::make_space_handle(dst_memory_space::name()),
```
because it will already check if the callback is set or not:
https://github.com/kokkos/kokkos/blob/b06886a3700f14c493df4ddbe3076501fd910be4/core/src/impl/Kokkos_Profiling.cpp#L429-L433
https://github.com/kokkos/kokkos/blob/b06886a3700f14c493df4ddbe3076501fd910be4/core/src/impl/Kokkos_Profiling.cpp#L279-L282


Again, it seems this is done all over the place so probably it's not to be addressed here, but should definitely be addressed.

Reply:
yeah we can do a follow up independent of this.